sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,131 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.json
+
+import org.apache.spark.sql.catalyst.{InternalRow, StructFilters}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.types.StructType
+
+/**
+ * The class provides API for applying pushed down source filters to rows with
+ * struct schema parsed from JSON records. It assumes that:
+ *   1. reset() is called before any skipRow() calls for new row.
+ *   2. skipRow() can be called for any valid index of the struct fields,
+ *      and in any order.
+ *
+ * @param filters The pushed down source filters. The filters should refer to
+ *                the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(filters: Seq[sources.Filter], schema: StructType)
+  extends StructFilters(filters, schema) {
+
+  /**
+   * Stateful JSON predicate that keeps track of its dependent references in the
+   * current row via `refCount`.
+   *
+   * @param predicate The predicate compiled from pushed down source filters.
+   * @param totalRefs The total amount of all filters references which the predicate
+   *                  compiled from.
+   * @param refCount The current number of predicate references in the row that have
+   *                 been not set yet. When `refCount` reaches zero, the predicate
+   *                 has all dependencies are set, and can be applied to the row.
+   */
+  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
+    def reset(): Unit = {
+      refCount = totalRefs
+    }
+  }
+
+  // Predicates compiled from the pushed down filters, and indexed by their references
+  // in the given schema. A predicate can occur in the array more than once if it
+  // has many references. For example:
+  //  schema: i INTEGER, s STRING
+  //  filters: IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc"))
+  //  predicates:
+  //    0: Array(IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  //    1: Array(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
```

Comment:
@MaxGekk, sorry if I rushed to read but do you mind elaborating this a bit more?

Reply:
I tried to explain the result step-by-step in the comments below. Could you point out particular step which is not clear for you.

Reply:
In the example, you can see per each field index, I form an array of predicates that have a reference to the field. This allows to consider only sub-set of predicates in the `skipRow()` call with particular index.
```
1: Array(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
```
`s` field has index `1`. `AlwaysTrue` filter doesn't have any reference by we should take it into account since we unpredictable order of JSON fields.  The `And(EqualTo("i", 0), StringStartsWith("s", "abc"))` has a reference to `s` in `StringStartsWith("s", "abc")`.  The filter `IsNotNull("i")` is not considered at all because it doesn't refer to the `s` field.